opt: crash in makeIndexConstraints for inverted indexes

The following statements crash the server:
```
CREATE TABLE j (a INT PRIMARY KEY, j JSON);
CREATE INVERTED INDEX ON j(j);
SELECT * FROM j WHERE j @> (NULL::JSON);

driver: bad connection
warning: connection lost!
```



```
E180327 03:52:03.336709 279 sql/conn_executor.go:514  [n1,client=127.0.0.1:54213,user=root] a SQL panic has occurred while executing "SELECT * FROM j WHERE j @> (NULL::JSON)": interface conversion: tree.Datum is tree.dNull, not *tree.DJSON
I180327 03:52:03.336743 279 sql/conn_executor.go:1816  [n1,client=127.0.0.1:54213,user=root] finishing connExecutor
E180327 03:52:03.336854 279 util/log/crash_reporting.go:185  [n1,client=127.0.0.1:54213,user=root] a panic has occurred!
E180327 03:52:03.338721 279 util/log/crash_reporting.go:451  [n1,client=127.0.0.1:54213,user=root] Reported as error c1c318536f7344da99a8af28808df7b7
panic: interface conversion: tree.Datum is tree.dNull, not *tree.DJSON [recovered]
        panic: panic while executing 1 statements: SELECT * FROM _ WHERE _ @> (_::JSON); caused by interface conversion: tree.Datum is tree.dNull, not *tree.DJSON

goroutine 279 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc425498000, 0x605d000, 0xc42081f240, 0x5aed280, 0xc42552c340)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:528 +0x365
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc425498000, 0x605d000, 0xc42081f240)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0x61
panic(0x5aed280, 0xc42552c340)
        /usr/local/Cellar/go/1.10/libexec/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint.(*indexConstraintCtx).makeInvertedIndexSpansForExpr(0xc424caf8a8, 0xc4254f0240, 0x3f00000003, 0xc424caf900, 0x606b001)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint/index_constraints.go:753 +0xb39
github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint.(*Instance).Init(0xc424caf8a8, 0xc4254f0240, 0x3f00000003, 0xc4203a2be0, 0x1, 0x1, 0x502c601, 0xc425498458, 0xc4203a2b20)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint/index_constraints.go:994 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.(*indexInfo).makeIndexConstraints(0xc424caf880, 0xc424b9ef50, 0xc4254f0240, 0x3f00000003, 0xc425498458, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_index_selection.go:468 +0x458
github.com/cockroachdb/cockroach/pkg/sql.(*planner).selectIndex(0xc4254983e8, 0x605d0c0, 0xc4253ba7e0, 0xc425308900, 0x0, 0x86e6100, 0x0, 0xc4204b0000, 0x481bd99, 0x8171b20)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_index_selection.go:155 +0x15aa
github.com/cockroachdb/cockroach/pkg/sql.expandScanNode(0x605d0c0, 0xc4253ba7e0, 0xc4254983e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0xc425308900, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:433 +0x10e
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x605d0c0, 0xc4253ba7e0, 0xc4254983e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x6053600, 0xc425308900, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:271 +0x1dab
github.com/cockroachdb/cockroach/pkg/sql.expandRenderNode(0x605d0c0, 0xc4253ba7e0, 0xc4254983e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0xc425386b00, 0xc425308900, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:446 +0x103
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x605d0c0, 0xc4253ba7e0, 0xc4254983e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x6053500, 0xc425386b00, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:274 +0x18d6
github.com/cockroachdb/cockroach/pkg/sql.(*planner).expandPlan(0xc4254983e8, 0x605d0c0, 0xc4253ba7e0, 0x6053500, 0xc425386b00, 0x6053500, 0xc425386b00, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:34 +0xce
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optimizePlan(0xc4254983e8, 0x605d0c0, 0xc4253ba7e0, 0x6053500, 0xc425386b00, 0xc420365522, 0x5, 0x5, 0x6053500, 0xc425386b00, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/optimize.go:47 +0xeb
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc4254983e8, 0x605d0c0, 0xc4253ba7e0, 0x605f980, 0xc42081f840, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:323 +0x2ea
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc425498000, 0x605d0c0, 0xc4253ba7e0, 0x605f980, 0xc42081f840, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:552 +0xb1b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc425498000, 0x605d0c0, 0xc4253ba7e0, 0x605f980, 0xc42081f840, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:369 +0xa8c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc425498000, 0x605d0c0, 0xc4253ba7e0, 0x605f980, 0xc42081f840, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:84 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc425498000, 0x605d000, 0xc42081f240, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:911 +0x1d2b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4204e21c0, 0x605d000, 0xc42081f240, 0x0, 0x0, 0xc425489009, 0x4, 0xc42548905c, 0x9, 0x6046440, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491 +0xf80
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc4204e21c0, 0x605d000, 0xc42081f240, 0xc42548e000, 0x5400, 0x15000, 0xc4200e9c90, 0xc420332b00, 0xc4203a94e0, 0xc420332af0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0xfc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf14
```

--- a/pkg/sql/logictest/testdata/logic_test/inverted_index
+++ b/pkg/sql/logictest/testdata/logic_test/inverted_index
@@ -161,6 +161,14 @@ index-join  0  index-join  ·      ·                                           
 ·           1  ·           table  d@primary                                    ·                ·
 
 query IT
+SELECT * from d where b @> NULL ORDER BY a;
+----
+
+query IT
+SELECT * from d where b @> (NULL::JSONB) ORDER BY a;
+----
+
+query IT
 SELECT * from d where b @>'{"a": "b"}' ORDER BY a;
 ----
 1  {"a": "b"}

--- a/pkg/sql/opt/idxconstraint/index_constraints.go
+++ b/pkg/sql/opt/idxconstraint/index_constraints.go
@@ -750,6 +750,12 @@ func (c *indexConstraintCtx) makeInvertedIndexSpansForExpr(
 		}
 
 		rightDatum := memo.ExtractConstDatum(rhs)
+
+		if rightDatum == tree.DNull {
+			c.contradiction(0 /* offset */, out)
+			return true
+		}
+
 		rd := rightDatum.(*tree.DJSON).JSON
 
 		switch rd.Type() {

